application/comit_node/src/http_api/routes/rfc003/swap_state.rs

```diff
@@ -194,38 +194,120 @@ impl<L: Ledger> From<rfc003::LedgerState<L>> for LedgerState<L::HtlcLocation, L:
 }
 
 impl SwapStatus {
-    pub fn new<AL: Ledger, BL: Ledger>(
-        swap_communication: &SwapCommunication<AL::Identity, BL::Identity>,
-        alpha_ledger: &LedgerState<AL::HtlcLocation, AL::Transaction>,
-        beta_ledger: &LedgerState<BL::HtlcLocation, BL::Transaction>,
+    pub fn new(
+        swap_communication_state: SwapCommunicationState,
+        alpha_ledger: rfc003::HtlcState,
+        beta_ledger: rfc003::HtlcState,
         error: &Option<rfc003::Error>,
     ) -> Self {
-        let swap_communication_state = &swap_communication.status;
-        let alpha_ledger = &alpha_ledger.status;
-        let beta_ledger = &beta_ledger.status;
-
         use self::SwapCommunicationState::*;
         use crate::swap_protocols::rfc003::HtlcState::*;
-        match (swap_communication_state, alpha_ledger, beta_ledger, error) {
-            (Rejected, _, _, None)
-            | (Accepted, Redeemed, Refunded, None)
-            | (Accepted, Refunded, Redeemed, None)
-            | (Accepted, Refunded, Refunded, None) => SwapStatus::NotSwapped,
-            (Accepted, Redeemed, Redeemed, None) => SwapStatus::Swapped,
-            (Sent, NotDeployed, NotDeployed, None) | (Accepted, _, _, None) => {
-                SwapStatus::InProgress
-            }
-            (swap_communication_state, alpha_ledger, beta_ledger, error) => {
-                log::warn!(
-                    "Internal failure with swap communication state {:?},\
-                     alpha ledger state {:?}, beta ledger state {:?} and error {:?}",
-                    swap_communication_state,
-                    alpha_ledger,
-                    beta_ledger,
-                    error
-                );
-                SwapStatus::InternalFailure
-            }
+
+        if let Some(e) = error {
+            log::debug!(target: "http-api", "derived SwapStatus is InternalFailure because: {:?}", e);
+            return SwapStatus::InternalFailure;
+        }
+
+        if swap_communication_state == Rejected {
+            return SwapStatus::NotSwapped;
+        }
+
+        match (alpha_ledger, beta_ledger) {
+            (Redeemed, Redeemed) => SwapStatus::Swapped,
+            (Refunded, _) | (_, Refunded) => SwapStatus::NotSwapped,
+            _ => SwapStatus::InProgress,
         }
     }
 }
+
+#[cfg(test)]
+impl quickcheck::Arbitrary for SwapCommunicationState {
```

Comment:
:+1: Omg yes quickcheck

Reply:
We should definitely make more use of quickcheck!